writers: Add writeFSharp and makeFSharpWriter functions to writers

moinessim opened the pull request:
Uses fsharp interactive (fsi) to run fsx script

Add tests

###### Motivation for this change

It would be awesome to leverage the scripting power of F# in the Nix ecosystem.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

moinessim commented:
@Lassulus @Profpatsch , what do you think?

Lassulus commented on pkgs/build-support/writers/default.nix:
```diff
@@ -301,6 +301,42 @@ let
     writePyPy3 "/bin/${name}";
 
+
+  makeFSharpWriter = { dotnet-sdk ? pkgs.dotnet-sdk, fsi-flags ? "", nugetDeps ? _: [] }: nameOrPath:
```
maybe we should name `nugetDeps` `libraries` like in the other writers? could be nice to have a common interface

moinessim replied:
Hi @Lassulus , thanks for taking a look.
Done.

Lassulus reviewed (approved):
looks good I guess. I'm no expert in F Sharp so not sure about most of it

Lassulus commented:
oh noes, wanted to merge but, now it has a merge conflict

Lassulus commented:
please rebase on master rather than merging it :)

Commit:
Add writeFSharp and makeFSharpWriter functions to writers

Uses fsharp interactive (fsi) to run fsx script

Expose mkNugetSource and mkNugetDeps functions

Use them in writers.writeFSharp and buildDotnetModule

Add tests

moinessim commented:
> please rebase on master rather than merging it :)

Done

IvarWithoutBones commented:
Thanks for the work! Could you ping me for future dotnet related stuff? I'd love to review it.

moinessim commented:
> Thanks for the work! Could you ping me for future dotnet related stuff? I'd love to review it before it gets merged

Sure! I'll try to remember... 😊

IvarWithoutBones commented:
Thanks!
